subsys/debug/tracing/cpu_stats.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_cpu_stats.h>
+#include <misc/printk.h>
+
+enum cpu_state {
+	CPU_STATE_IDLE,
+	CPU_STATE_NON_IDLE,
+	CPU_STATE_SCHEDULER
+};
+
+static enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
+static enum cpu_state cpu_state_before_interrupts;
+
+static u32_t last_time;
+static u64_t idle_cnt;
+static u64_t non_idle_cnt;
+static u64_t sched_cnt;
+static int nested_interrupts;
+static struct k_thread *current_thread;
+
+#ifndef CONFIG_SMP
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
+{
+#ifdef CONFIG_SMP
+	return thread->base.is_idle;
+#else
+	return thread == _idle_thread;
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
+{
+	u32_t time = k_cycle_get_32();
+
+	if (time >= last_time) {
+		(*cnt) += (time - last_time);
+	} else {
+		(*cnt) += (UINT32_MAX - last_time + 1 + time);
+	}
+	last_time = time;
+}
+
+static void cpu_stats_update_counters(void)
+{
+	switch (last_cpu_state) {
+	case CPU_STATE_IDLE:
+		increase_counter(&idle_cnt);
+		break;
+
+	case CPU_STATE_NON_IDLE:
+		increase_counter(&non_idle_cnt);
+		break;
+
+	case CPU_STATE_SCHEDULER:
+		increase_counter(&sched_cnt);
+		break;
+
+	default:
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+}
+
+u64_t cpu_stats_sched_get_ns(void)
```

Comment:
Could we replace these function by `cpu_stats_get(struct cpu_stats *stats)`?
IMHO such approach will be simpler and easier to extend.

Reply:
done